keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,360 @@
+
+Individual kubernetes vendors can easily customize their deployments and still use upstream Cluster Autoscaler and other ecosystem components that care about node health with this design.
+
+
+#### Optional Extension - Track state of components along with Health
+
+Each of the individual node extension will most likely want to expose domain specific status for instrospection purposes.
+This status can be expressed as plugin specific CRDs which may be a good starting point.
+Down the line though, having a generic abstraction(s) to represent the Status of different plugins or extensions will be necessary to empower the wider introspection & remediation ecosystem.
+
+Assuming it is possible to define Status for some of the popular plugins like CSI, device plugins, kube proxy, logging & monitoring agents, the Lease API based approach can be extended as follows to easily track status of extensions in addition to Health.
+
+1. Embed the "Lease" API object in an extension specific Status object. An alternative is to have a local Lease object with the same name as the Status object.
+2. Similar to the workflow described above, the node controller, kubelet or extensions will register their status object as an extension via the Node.Spec.SystemExtensions field.
+3. The Node Controller will be updated to extract the Lease object (or identifying a separate Lease object by name in the local namespace) and influence health of extensions similar to the workflow described above.
+4. Ecosystem components like Monitoring, repair/remediation systems will be able to deterministically identify and expose or act on the status of the extensions.
+
+Here are some API illustations.
+
+##### Node Object extension
+
+```yaml
+type ExtensionFoo struct {
+ TypeMeta
+ ObjectMeta
+ Spec ExtensionFooStatus
+}
+
+type ExtensionFooStatus {
+  v1beta1.LeaseSpec
+  OtherFooStatus interface{}
+}
+
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
+...
+}
+
+type SystemExtension struct {
+   Reference ObjectReference
+   AffectsReadiness bool // Ignore non-critical node extensions from node readiness
+}
+```
+
+##### Sample Extension Status Object
+
+```yaml
+type DevicePluginExtension struct {
+ TypeMeta
+ ObjectMeta
+ Status DevicePluginExtensionStatus
+}
+
+type DevicePluginExtensionStatus struct {
+  LeaseSpec
+  ComputeResource struct{ // embedded for brevity
+    Name: “foo.com/bar”
+    Quantity: x // Sample fields for illustration purposes
+    UnHealthy: y
+    DeviceIDs []string
+  }
+}
+
+```
+
+One of the goals for this proposal is to evaluate the viability and usefullness of combining Status and Health for extensions.
+
+### Approach B - Use Readiness of Node extension pods (Preferred)
+
+This approach assumes that all node level extensions will get deployed as kubernetes pods and that the health of those extension pods can be exposed via existing Readiness probes.
```

Comment:
So it sounds like the model envisaged is pod-per-Node.  I do think the selector would enable selection of a shared pod, but it isn't clear that a pod could express that some subset of nodes are bad and some are good.

I think anything that programs infrastructure would end up needing a "proxy pod" per node?

Reply:
Yes, in an ideal world all critical system daemons would be run as pods (including k'let :) ). Until that happens, there can be placeholder pods that reflect the health of system daemons not run as pods for the purposes of monitoring/introspection.